### PR TITLE
Fixes #11130: improve rudder-tests techniques scenario

### PR DESCRIPTION
**Pull request by Fdall**
https://www.rudder-project.org/redmine/issues/11130

### CODE REVIEW

File: scenario/lib.py [hunk 1]
```diff
@@ -234,11 +234,12 @@ def get_tests():
       #                  "init": "init command",  # it must be relative since the PATH will include the metadata file path
```
Comment (peckpeck) [inline scenario/lib.py]: it is now inits
try to fix little mistakes as soon as you see them;

File: rtf [hunk 2]
```diff
@@ -1217,6 +1217,8 @@ if __name__ == "__main__":
       filter = args['--filter']
       if filter == []:
         filter = None
+# add the reset scenario execution before a technique test
```
Comment (peckpeck) [inline rtf]: comment indent

File: rtf [hunk 3]
```diff
@@ -1217,6 +1217,8 @@ if __name__ == "__main__":
...
+      platform.run_scenario('reset', args['--format'], not args['--no-finally'], args['--stop'], filter, client_path, params)
```
Comment (peckpeck) [inline rtf]: 👍

Comment (peckpeck) [inline rtf]: shouldn't you update reset too, to remove directives ?

File: scenario/lib.py [hunk 4]
```diff
@@ -234,11 +234,12 @@ def get_tests():
...
       #                  "directives": [ "directive_file", ... ],
       #                  "checks": [ "check_script.rb", ... ],
+      #                  "sharedFiles": ["file to upload", ...],
       #                  "compliance": "expected compliance" }
       for metadata in metadatas:
         root = os.path.abspath(os.path.dirname(metadata_file))
-        metadata['local_root'] = root
-        metadata['remote_root'] = root.replace(technique_root,  "/var/rudder/configuration-repository/techniques/")
+        metadata[u'local_root'] = root
+        metadata[u'remote_root'] = root.replace(technique_root,  "/var/rudder/configuration-repository/techniques/")
```
Comment (peckpeck) [inline scenario/lib.py]: why did you add the u ?

File: scenario/lib.py [hunk 5]
```diff
@@ -264,7 +265,13 @@ def get_tests():
           inits.append(path)        
         metadata['inits'] = inits
 
+        # make sharedFiles path absolute
+        sharedFiles = []
+        for iFile in metadata['sharedFiles']:
+          _file_must_exist(iFile)
```
Comment (peckpeck) [inline scenario/lib.py]: i'd prefer all files to be relative, so prepend the root

File: scenario/technique.py [hunk 6]
```diff
@@ -33,12 +34,18 @@
...
   # Run init script
   if 'inits' in test:
     run_on("agent", 'techniques/technique_init', Err.BREAK, INITS=",".join(test['inits']), SERVER_VERSION=scenario.server_rudder_version())
 
+  # Add the shared-Files on the server
+  if 'sharedFiles' in test:
+    for iFile in test['sharedFiles']:
+        basename = os.path.basename(iFile)
+        run_on("server", 'techniques/technique_uploadFiles', Err.BREAK, FILE=iFile, BASENAME=basename)
+        run_on("server", 'techniques/technique_sharedFiles', Err.BREAK, BASENAME=basename)
```
Comment (peckpeck) [inline scenario/technique.py]: contrary to what i said earlier, you could merge both

File: spec/tests/techniques/technique_sharedFiles.rb [hunk 7]
```diff
@@ -0,0 +1,11 @@
+require 'spec_helper'
+basename = $params['BASENAME']
```
Comment (peckpeck) [inline spec/tests/techniques/technique_sharedFiles.rb]: You could guess basename in ruby, example https://stackoverflow.com/questions/374326/how-to-get-filename-without-extension-from-file-path-in-ruby

File: rtf (hunk 3 above)
Comment (Fdall) [inline rtf]: I was mistaking, it is already the case.

**Comment by Fdall** [comment]
Commit modified

**Comment by Fdall** [comment]
Commit modified

File: scenario/lib.py [hunk 8]
```diff
@@ -231,9 +231,10 @@ def get_tests():
       metadatas = json.load(fd)
       # metadata content : [ { test }, ... ]
       # test content : { "name": "test_name",
-      #                  "init": "init command",  # it must be relative since the PATH will include the metadata file path
+      #                  "inits": "inits command",  # it must be relative since the PATH will include the metadata file path
```
Comment (peckpeck) [inline scenario/lib.py]: between [] like below

File: spec/tests/techniques/technique_sharedFiles.rb [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+file = $params['FILE']
+basename = File.basename(file)
+destination = "/var/rudder/configuration-repository/shared-files"
+
+describe "File Upload" do
+  send_file(file, "/tmp/" + basename)
+end
+
+describe command("chown root /tmp/" + basename + "; cp /tmp/" + basename + " " + destination + "/") do
```
Comment (peckpeck) [inline spec/tests/techniques/technique_sharedFiles.rb]: cp + rm = mv
use chown -r just in case we are copying a directory

File: spec/tests/user_rule.rb [hunk 10]
```diff
@@ -9,7 +9,7 @@
 describe "Add a user directive and a rule"  do
 
   # Add a User directive
-  describe command($rudderCli + " directive create --json=" + directiveFile + " userManagement user") do
+  describe command($rudderCli + " directive create --json=" + directiveFile + " userManagement user | jq '.directives | .[0].id'") do
```
Comment (peckpeck) [inline spec/tests/user_rule.rb]: .directives[0].id

**Comment by Fdall** [comment]
Commit modified

File: spec/tests/techniques/technique_sharedFiles.rb [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+
+file = $params['FILE']
+basename = File.basename(file)
+destination = "/var/rudder/configuration-repository/shared-files"
+
+describe "File Upload" do
+  send_file(file, "/tmp/" + basename)
+end
+
+describe command("chown -r root /tmp/" + basename + "; mv /tmp/" + basename + " " + destination + "/") do
```
Comment (peckpeck) [inline spec/tests/techniques/technique_sharedFiles.rb]: this doesn't test if chown returned true, you can either split this in 2 tests or use '&&' instead of ';'

**Comment by Fdall** [comment]
Commit modified

**Comment by Fdall** [comment]
Commit modified

File: rtf [hunk 12]
```diff
@@ -354,11 +354,11 @@ class Vagrant(Host):
 
   def push_techniques(self, directory):
     """ Replace all techniques with the provided ones """
-    # First, empty techniques directory
-    self.run("cd /var/rudder/configuration-repository/techniques/ && find . -mindepth 1 -maxdepth 1 -type d -not -name system | xargs rm -rf ", fail_exit=False)
+    # First, empty techniques directory matching the tested one
+    name = os.path.basename(os.path.abspath(directory))
+    self.run("cd /var/rudder/configuration-repository/techniques/ && find -mindepth 1 -maxdepth 2 -type d -name " + name + " -not -name system | xargs rm -rf ", fail_exit=False)
```
Comment (peckpeck) [inline rtf]: Why maxdepth 2 and not more ?
-not -name system  is not useful anymore

File: scenario/technique.py [hunk 13]
```diff
@@ -33,12 +33,17 @@
...
     rule_name=test['name']
   else:
-    rule_name="test_"+str(test_idi)
+    rule_name="test_"+str(test_id)
 
   # Run init script
   if 'inits' in test:
-    run_on("agent", 'techniques/technique_init', Err.BREAK, INITS=",".join(test['inits']), SERVER_VERSION=scenario.server_rudder_version())
+    for iInit in test['inits']:
+        run_on("agent", 'techniques/technique_init', Err.BREAK, INITS=iInit, SERVER_VERSION=scenario.server_rudder_version())
```
Comment (peckpeck) [inline scenario/technique.py]: INIT without S

**Comment by Fdall** [comment]
Commit modified

**Comment by Fdall** [comment]
Commit modified

**Comment by Fdall** [comment]
Commit modified

File: scenario/lib.py [hunk 14]
```diff
@@ -264,7 +265,14 @@ def get_tests():
           inits.append(path)        
         metadata['inits'] = inits
 
+        # make sharedFiles path absolute
+        sharedFiles = []
+        for iFile in metadata['sharedFiles']:
+            path = root+'/'+iFile
```
Comment (peckpeck) [inline scenario/lib.py]: Even if it's not recommended for python, we use 2 spaces indent in rudder

File: rtf (hunk 12 above)
Comment (peckpeck) [inline rtf]: You still forgot this

File: scenario/technique.py [hunk 15]
```diff
@@ -33,12 +33,18 @@
...
   # Run init script
   if 'inits' in test:
-    run_on("agent", 'techniques/technique_init', Err.BREAK, INITS=",".join(test['inits']), SERVER_VERSION=scenario.server_rudder_version())
+    for iInit in test['inits']:
+        run_on("agent", 'techniques/technique_init', Err.BREAK, INIT=iInit, SERVER_VERSION=scenario.server_rudder_version())
 
+  # Add the shared-Files on the server
+  if 'sharedFiles' in test:
+    for iFile in test['sharedFiles']:
+        print(iFile)
```
Comment (peckpeck) [inline scenario/technique.py]: This looks like a debug print

Comment (Fdall) [inline scenario/technique.py]: Fixed

File: rtf (hunk 12 above)
Comment (Fdall) [inline rtf]: Fixed

File: scenario/lib.py (hunk 14 above)
Comment (Fdall) [inline scenario/lib.py]: Fixed

**Comment by Fdall** [comment]
Commit modified
